### PR TITLE
copr/mysql/json: support json_keys

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Wangweizhen <hawking.rei@gmail.com>
###  What have you changed?

Make JSON support the json_keys. It's the implementation for JSON_KEYS in Mysql.

###  What is the type of the changes?
- New feature

###  How is the PR tested?
- Unit test

###  Does this PR affect tidb-ansible?

- No.

### CODE REVIEW

**Comment by hawkingrei** [comment]
@lonng @breeswish PTAL

**Comment by hawkingrei** [comment]
@lonng @breeswish PTAL

**Comment by mapleFU** [comment]
Would you mind implement `json_keys` in  `builtin_json.rs` or `impl_json.rs`?

File: components/tidb_query/src/codec/mysql/json/json_keys.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            };
+        }
+        Ok(json_keys(self))
+    }
+}
+
+fn json_keys(j: &Json) -> Option<Json> {
+    match *j {
+        Json::Object(ref map) => {
+            let mut keys: Vec<Json> = Vec::new();
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/json/json_keys.rs]: Prefer `map.keys().map(|key| Json::String(key.to_string())).collect()`

File: components/tidb_query/src/codec/mysql/json/json_keys.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+use super::super::Result;
+use super::path_expr::PathExpression;
+use super::Json;
+
+impl Json {
+    pub fn keys(&self, path_expr_list: &[PathExpression]) -> Result<Option<Json>> {
+        if !path_expr_list.is_empty() {
+            if path_expr_list.len() > 1 {
+                return Err(box_err!("Incorrect parameter count"));
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/json/json_keys.rs]: Better to contain the `path_expr_list ` to error message to help locate problem.

File: components/tidb_query/src/codec/mysql/json/json_keys.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            }
+            for expr in path_expr_list {
+                if expr.contains_any_asterisk() {
+                    return Err(box_err!("Invalid path expression"));
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/json/json_keys.rs]: ditto

File: components/tidb_query/src/codec/mysql/json/json_keys.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                }
+            }
+            let json_option = self.extract(path_expr_list);
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/json/json_keys.rs]: `return Ok(self.extract(path_expr_list).map(|j| json_keys(&j)))`

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 2 above)
Comment (mapleFU) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: This error is an `Incorrect parameter count`, may be we just need to show the length of 
 `path_expr_list`?

**Comment by hawkingrei** [comment]
@mapleFU Yes,I will

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 2 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: > Would you mind implement `json_keys` in `builtin_json.rs` or `impl_json.rs`?

Yes,I will

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 4 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: done

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 3 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: done

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 2 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: done

File: components/tidb_query/src/codec/mysql/json/json_keys.rs (hunk 1 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_keys.rs]: done

File: components/tidb_query/src/codec/mysql/json/json_keys.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+use super::Json;
+
+impl Json {
+    pub fn keys(&self, path_expr_list: &[PathExpression]) -> Result<Option<Json>> {
+        if !path_expr_list.is_empty() {
+            if path_expr_list.len() > 1 {
+                return Err(box_err!("Incorrect parameter: {:?}", path_expr_list));
+            }
+            for expr in path_expr_list {
+                if expr.contains_any_asterisk() {
```
Comment (mapleFU) [inline components/tidb_query/src/codec/mysql/json/json_keys.rs]: You can use 
```
if path_expr_list.iter().any(|expr| expr.contains_any_asterisk()) {
    return Err(box_err!("Invalid path expression: {:?}", path_expr_list));
}
```

**Comment by mapleFU** [comment]
The rest LGTM.

**Comment by mapleFU** [comment]
Would you mind implement `json_keys` in `builtin_json.rs` or `impl_json.rs` and open push down tests for it? Since we can using integration test to test it.

**Comment by hawkingrei** [comment]
@iosmanthus PTAL

**Comment by breeswish** [comment]
@iosmanthus PTAL, thanks!

**Pull request merged**